src/Disks/ObjectStorages/DiskObjectStorageTransaction.cpp

```diff
@@ -1177,25 +1210,69 @@ TransactionCommitOutcomeVariant DiskObjectStorageTransaction::tryCommit(const Tr
         throw;
     }
 
+    /// after successful commit of metadata transaction we can not rollback transaction
+
     StoredObjects objects_to_remove;
     for (const auto & operation : operations_to_execute)
-        operation->finalize(objects_to_remove);
+    {
+        try
+        {
+            operation->finalize(objects_to_remove);
+        }
+        catch (...)
+        {
+            tryLogCurrentException(
+                __PRETTY_FUNCTION__,
+                fmt::format("An error occurred while finalizing transaction's operation ({})", operation->getInfoForLog()));
+        }
+    }
+
+    try
+    {
+        object_storage.removeObjectsIfExist(objects_to_remove);
+    }
+    catch (...)
+    {
+        tryLogCurrentException(
+            __PRETTY_FUNCTION__,
+            "An error occurred while removing objects during finalizing transaction");
+    }
 
-    object_storage.removeObjectsIfExist(objects_to_remove);
     operations_to_execute.clear();
     is_committed = true;
     return outcome;
 }
 
 void DiskObjectStorageTransaction::undo()
```

Comment:
Let's change the api here. What if the function will return `std::expected<void, std::exception_ptr>` and will be noexcept?